### PR TITLE
Command design pattern

### PR DESCRIPTION
**Pull request by tomvai15**
Implemented command. Supported commands: undo, place, upgrade, remove.

Item level is show next to it.
image

### CODE REVIEW

File: src/TowerDefense.Api/Battle/Command/CommandExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public interface ICommandExecutor
+    {
+        void Execute(Command command);
+        public void Undo();
+    }
+    public class CommandExecutor : ICommandExecutor
+    {
+        private readonly CommandHistory _commandHistory = new CommandHistory();
+        private readonly IPlayer player;
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Command/CommandExecutor.cs:13]: Add a prefix of _ to player field https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#camel-case

File: src/TowerDefense.Api/Battle/Command/CommandExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        public CommandExecutor(IPlayer player)
+        {
+            this.player = player;
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Command/CommandExecutor.cs:17]: avoid using this

File: src/TowerDefense.Api/Battle/Command/CommandExecutor.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public void Execute(Command command)
+        {
+            command.Execute(player);
+            if (command is IRevertable revertable)
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Command/CommandExecutor.cs:23]: reduce nesting by inverting if

File: src/TowerDefense.Api/Battle/Command/CommandHistory.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿namespace TowerDefense.Api.Battle.Command
+{
+    public class CommandHistory
+    {
+        private readonly Stack<IRevertable> _executedCommands = new Stack<IRevertable>();
+        public void Push(IRevertable command)
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Command/CommandHistory.cs:6]: add spaces in between the methods

File: src/TowerDefense.Api/Battle/Command/PlaceCommand.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+﻿using TowerDefense.Api.Battle.Grid;
+using TowerDefense.Api.Models;
+using TowerDefense.Api.Models.Items;
+using TowerDefense.Api.Models.Player;
+
+namespace TowerDefense.Api.Battle.Command
+{
+    public class PlaceCommand : Command, IRevertable
+    {
+        private readonly string inventoryItemId;
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Command/PlaceCommand.cs:10]: prefix _

File: src/TowerDefense.Api/Battle/Command/RemoveCommand.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+﻿using TowerDefense.Api.Models.Items;
+using TowerDefense.Api.Models.Player;
+
+namespace TowerDefense.Api.Battle.Command
+{
+    public class RemoveCommand : Command, IRevertable
+    {
+        private readonly int gridItemId;
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Command/RemoveCommand.cs:8]: prefix _

File: src/TowerDefense.Api/Battle/Command/UpgradeCommand.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            this.gridItemId = gridItemId;
+        }
+
+        public override void Execute(IPlayer player)
+        {
+            var gridItem = player.ArenaGrid.GridItems[gridItemId];
+            previousItem = gridItem.Item;
+
+            IItem upgradedItem;
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Command/UpgradeCommand.cs:23]: consider using new syntax. Check out the link https://github.com/antanasramana/TowerDefense/blob/f1d30971d5705d862ee4fd9acf903c3daecfdd84/src/TowerDefense.Api/Controllers/PlayerController.cs#L35

File: src/TowerDefense.Api/Battle/Handlers/BattleHandler.cs [hunk 8]
```diff
@@ -46,8 +46,8 @@ public void HandleEndTurn(string playerName)
             player2.Money += PlayerEarnedMoneyAfterAttack(player2AttackDeclarations);
 
             // Notify opposing players grid items to receive attack
-            var player1AttackResults = _gameState.GridPublishers[0].Notify(player1AttackDeclarations);
-            var player2AttackResults = _gameState.GridPublishers[1].Notify(player2AttackDeclarations);
+            var player1AttackResults = player2.Publisher.Notify(player1AttackDeclarations);
```
Comment (antanasramana) [inline src/TowerDefense.Api/Battle/Handlers/BattleHandler.cs:49]: good job

File: src/TowerDefense.Api/Decorator/BaseDecorator.cs [hunk 9]
```diff
@@ -21,6 +22,7 @@ public BaseDecorator(IItem item)
             _item = item;
             Id = item.Id;
             Price = item.Price;
+            Level = item.Level+1;
```
Comment (antanasramana) [inline src/TowerDefense.Api/Decorator/BaseDecorator.cs:25]: spaces in between operands

File: src/TowerDefense.Api/Battle/Command/CommandExecutor.cs (hunk 3 above)
Comment (tomvai15) [reply src/TowerDefense.Api/Battle/Command/CommandExecutor.cs:23]: Bruh

**Review by EimantasMaz — APPROVED** [review]
Antanas comments seem to be resolved

**Pull request merged**